packages/protobuf/src/lib.ts

```diff
@@ -159,7 +171,46 @@ export const TypeSpecProtobufLibrary = createTypeSpecLibrary({
   emitter: { options: EmitterOptionsSchema },
 });
 
-export const { reportDiagnostic } = TypeSpecProtobufLibrary;
+export type Diagnostic = Parameters<typeof TypeSpecProtobufLibrary.reportDiagnostic>[1];
+
+const __DIAGNOSTIC_CACHE = new WeakMap<Program, Map<DiagnosticTarget, Set<Diagnostic["code"]>>>();
+
+function getDiagnosticCache(program: Program) {
+  let cache = __DIAGNOSTIC_CACHE.get(program);
+  if (!cache) {
+    cache = new Map();
+    __DIAGNOSTIC_CACHE.set(program, cache);
+  }
+  return cache;
+}
+
+function getAppliedCodesForTarget(program: Program, target: DiagnosticTarget) {
+  const cache = getDiagnosticCache(program);
+  let codes = cache.get(target);
+  if (!codes) {
+    codes = new Set();
+    cache.set(target, codes);
+  }
+  return codes;
+}
+
+export const reportDiagnostic = Object.assign(TypeSpecProtobufLibrary.reportDiagnostic, {
+  /**
+   * Report a TypeSpec protobuf diagnostic, but only once per target per diagnostic code.
+   *
+   * This is useful in situations where a function that reports a recoverable diagnostic may be called multiple times.
+   */
+  once: function (program: Program, diagnostic: Diagnostic & { target: DiagnosticTarget }) {
```

Comment:
intresting idea, we talked about deduplicating diagnostic a while back where it would just do automatically but this turned out a little harder but if you can be explicit in some case it could be nice to have built in